File: Aniket_Code.js
class DiscountRule {
    constructor(productQuantityDetails) {
        this.productQuantityDetails = productQuantityDetails;
        this.flat10DisVal = 200;
        this.bulk5DisVal = 10;
        this.bulk10DisReq = 20;
        this.tiered50DisReq = 30;
        this.tiered50DisReq2 = 15;
        this.shippingLot = 10;
        this.shippingCostPerUnit = 5;
    }

    getWrapCost() {
        let totalGiftWrapCost = 0;
        for (let sProduct in this.productQuantityDetails) {
            if (this.productQuantityDetails[sProduct]["gift_wrap"] === "y") {
                totalGiftWrapCost += this.productQuantityDetails[sProduct]["quantity"] * 1;
            }
        }
        return totalGiftWrapCost;
    }

    getTotalSum() {
        let totalSum = 0;
        for (let sProduct in this.productQuantityDetails) {
            totalSum += this.productQuantityDetails[sProduct]["price"] * this.productQuantityDetails[sProduct]["quantity"];
        }
        return totalSum;
    }

    getTotalQty() {
        let totalQty = 0;
        for (let sProduct in this.productQuantityDetails) {
            totalQty += this.productQuantityDetails[sProduct]["quantity"];
        }
        return totalQty;
    }

    getShippingCost() {
        const totalAllQuantity = this.getTotalQty();
        let totalShipmentLot = Math.floor(totalAllQuantity / this.shippingLot);
        if (totalAllQuantity % this.shippingLot !== 0) {
            totalShipmentLot++;
        }
        return totalShipmentLot * this.shippingCostPerUnit;
    }

    flat10Discount() {
        const discountPrice = 10;
        if (this.getTotalSum() > this.flat10DisVal) {
            const afterDiscountTotal = this.getTotalSum() - discountPrice;
            return ["flat_10_discount", "Yes", afterDiscountTotal, this.getTotalSum() - afterDiscountTotal];
        }
    }

    bulk5Discount() {
        let totalSum = 0;
        const discountPer = 5;
        let discountStatus = "No";
        for (let sProduct in this.productQuantityDetails) {
            if (this.productQuantityDetails[sProduct]["quantity"] > this.bulk5DisVal) {
                totalSum += this.productQuantityDetails[sProduct]["price"] *
                    this.productQuantityDetails[sProduct]["quantity"] * ((100 - discountPer) / 100);
                discountStatus = "Yes";
            } else {
                totalSum += this.productQuantityDetails[sProduct]["price"] * this.productQuantityDetails[sProduct]["quantity"];
            }
        }
        return ["bulk_5_discount", discountStatus, totalSum, this.getTotalSum() - totalSum];
    }

    bulk10Discount() {
        let totalSum = this.getTotalSum();
        const discountPer = 10;
        if (this.getTotalQty() > this.bulk10DisReq) {
            totalSum *= (100 - discountPer) / 100;
            return ["bulk_10_discount", "Yes", totalSum, this.getTotalSum() - totalSum];
        } else {
            return ["bulk_10_discount", "No", totalSum, this.getTotalSum() - totalSum];
        }
    }

    tiered50Discount() {
        const totalCount = this.getTotalQty();
        let totalSum = 0;
        const discountPer = 50;
        let discountStatus = "No";
        for (let sProduct in this.productQuantityDetails) {
            if (totalCount > this.tiered50DisReq && this.productQuantityDetails[sProduct]["quantity"] > this.tiered50DisReq2) {
                totalSum += this.tiered50DisReq2 * this.productQuantityDetails[sProduct]["price"];
                totalSum += ((this.productQuantityDetails[sProduct]["quantity"] - this.tiered50DisReq2) *
                    this.productQuantityDetails[sProduct]["price"] * ((100 - discountPer) / 100));
                discountStatus = "Yes";
            } else {
                totalSum += this.productQuantityDetails[sProduct]["price"] * this.productQuantityDetails[sProduct]["quantity"];
            }
        }
        return ["tiered_50_discount", discountStatus, totalSum, this.getTotalSum() - totalSum];
    }
}

function takeUserInput(productCat) {
    const finalProductOrder = { ...productCat };
    for (let sProduct in productCat) {
        const productQuantity = parseInt(prompt(`Please Enter ${sProduct} quantity You Want?`));
        const giftWrap = prompt("Want to wrap Product as Gift? Enter 'y' or 'n'");
        finalProductOrder[sProduct]["quantity"] = productQuantity;
        finalProductOrder[sProduct]["gift_wrap"] = giftWrap;
    }
    return finalProductOrder;
}

const productCatalogue = takeUserInput({
    "Product_A": { "price": 20, "quantity": 0 },
    "Product_B": { "price": 40, "quantity": 0 },
    "Product_C": { "price": 50, "quantity": 0 }
});

const getDiscountRule = new DiscountRule(productCatalogue);
const allDiscountDetails = [];
allDiscountDetails.push(getDiscountRule.flat10Discount());
allDiscountDetails.push(getDiscountRule.bulk5Discount());
allDiscountDetails.push(getDiscountRule.bulk10Discount());
allDiscountDetails.push(getDiscountRule.tiered50Discount());

const totalAmountBeforeDiscount = getDiscountRule.getTotalSum();
const bestDiscount = allDiscountDetails.reduce((maxDiscount, discount) => (discount && discount[3] > maxDiscount[3] ? discount : maxDiscount), [0, "", 0, 0]);
const finalAmountAfterDiscount = bestDiscount[2];
const finalDiscountAmount = bestDiscount[3];
const finalDiscountName = bestDiscount[0];
const giftWrappingCost = getDiscountRule.getWrapCost();
const totalShippingCost = getDiscountRule.getShippingCost();
const finalTotalBillAmount = finalAmountAfterDiscount + giftWrappingCost + totalShippingCost;

console.log(`######### FINAL OUTPUT #########`);
for (let sProductDet in productCatalogue) {
    console.log(`${sProductDet}, Quantity: ${productCatalogue[sProductDet]['quantity']}, Total Amount: ${productCatalogue[sProductDet]['quantity'] * productCatalogue[sProductDet]['price']}`);
}
console.log(`Total Amount Before Discount: ${totalAmountBeforeDiscount}`);
console.log(`Discount Name: ${finalDiscountName}, Discount Amount: ${finalDiscountAmount}`);
console.log(`Gift Wrap Fee: ${giftWrappingCost}, Shipping Fee: ${totalShippingCost}`);
console.log(`Total: ${finalTotalBillAmount}`);

File: Aniket_Code.py
class Discount_rule:
    def __init__(self,product_qnty_details):
        self.product_qnty_details=product_qnty_details
        self.flat_10_dis_val=200
        self.bulk_5_dis_val=10
        self.bulk_10_dis_req=20
        self.tiered_50_dis_req=30
        self.tiered_50_dis_req_2=15
        self.shipping_lot=10
        self.shipping_cost_per_unir=5
        
    def to_get_wrap_cost(self):
        total_gify_wrap_cost=0
        for s_product in self.product_qnty_details:
            if self.product_qnty_details[s_product]["gift_wrap"]=="y":
                total_gify_wrap_cost=total_gify_wrap_cost+self.product_qnty_details[s_product]["quantity"]*1
        return total_gify_wrap_cost
                 
    def to_get_total_sum(self):
        total_sum=0
        for s_product in self.product_qnty_details:
            total_sum=total_sum+(self.product_qnty_details[s_product]["price"]*self.product_qnty_details[s_product]["quantity"])
        return total_sum

    def to_get_total_qty(self):
        total_qnty=0
        for s_product in self.product_qnty_details:
            total_qnty=total_qnty+self.product_qnty_details[s_product]["quantity"]
        return total_qnty
    
    def to_get_shipping_cost(self):
        total_all_quantity=self.to_get_total_qty()
        total_shipment_lot=total_all_quantity//self.shipping_lot
        if total_all_quantity%self.shipping_lot!=0:
            total_shipment_lot=total_shipment_lot+1
        return total_shipment_lot*self.shipping_cost_per_unir
        
    def flat_10_discount(self):
        discount_price=10
        if self.to_get_total_sum() > self.flat_10_dis_val:
            after_discount_total=self.to_get_total_sum()-discount_price
            return ["flat_10_discount","Yes",after_discount_total,self.to_get_total_sum()-after_discount_total]
        
    def bulk_5_discount(self):
        total_sum=0
        discount_per=5
        discount_status="No"
        for s_product in self.product_qnty_details:
            if self.product_qnty_details[s_product]["quantity"]>self.bulk_5_dis_val:
                total_sum=total_sum+(self.product_qnty_details[s_product]["price"]*self.product_qnty_details[s_product]["quantity"]
                                    *((100-discount_per)/100))
                discount_status="Yes"
            else:
                total_sum=total_sum+(self.product_qnty_details[s_product]["price"]*self.product_qnty_details[s_product]["quantity"])
        return ["bulk_5_discount",discount_status,total_sum,self.to_get_total_sum()-total_sum]
    
    def bulk_10_discount(self):
        total_sum=self.to_get_total_sum()
        discount_per=10
        if self.to_get_total_qty()> self.bulk_10_dis_req:
            total_sum=total_sum*((100-discount_per)/100)
            return ["bulk_10_discount","Yes",total_sum,self.to_get_total_sum()-total_sum]
        else:
            return ["bulk_10_discount","No",total_sum,self.to_get_total_sum()-total_sum]
        
    def tiered_50_discount(self):
        total_count=self.to_get_total_qty()
        total_sum=0
        discount_per=50
        discount_status="No"
        for s_product in self.product_qnty_details:
            if total_count>self.tiered_50_dis_req and self.product_qnty_details[s_product]["quantity"]>self.tiered_50_dis_req_2:
                total_sum=total_sum+(self.tiered_50_dis_req_2*self.product_qnty_details[s_product]["price"])
                total_sum=total_sum+((self.product_qnty_details[s_product]["quantity"]-self.tiered_50_dis_req_2)*
                          self.product_qnty_details[s_product]["price"]*((100-discount_per)/100))
                discount_status="Yes"
            else:
                total_sum=total_sum+(self.product_qnty_details[s_product]["price"]*self.product_qnty_details[s_product]["quantity"])
        
        return ["tiered_50_discount",discount_status,total_sum,self.to_get_total_sum()-total_sum]
    
            
def take_user_input(product_cat):
    final_product_order=product_cat.copy()
    for s_product in product_cat:
        product_quantity=eval(input(f"Please Enter {s_product} quantity You Want?"))
        gift_wrap=input(f"Want to wrap Product as Gift? 'Enter y or n'")
        final_product_order[s_product]["quantity"]=product_quantity
        final_product_order[s_product]["gift_wrap"]=gift_wrap
    return final_product_order

product_catalogue=take_user_input({
    "Product_A" :{"price":20,
                 "quantity":0},
    "Product_B" :{"price":40,
                 "quantity":0},
    "Product_C" :{"price":50,
                 "quantity":0}
})

get_discount_rule=Discount_rule(product_catalogue)
all_discount_details=[]
all_discount_details.append(get_discount_rule.flat_10_discount())
all_discount_details.append(get_discount_rule.bulk_5_discount())
all_discount_details.append(get_discount_rule.bulk_10_discount())
all_discount_details.append(get_discount_rule.tiered_50_discount())

total_amount_before_dic=get_discount_rule.to_get_total_sum()        
best_discount=max(all_discount_details, key=lambda x: x[3])
final_amount_after_discount = best_discount[2]
final_discount_amount = best_discount[3]
final_discount_name = best_discount[0]
gift_wrapping_cost=get_discount_rule.to_get_wrap_cost()
total_shipping_cost=get_discount_rule.to_get_shipping_cost()
Final_total_bill_amount=final_amount_after_discount+gift_wrapping_cost+total_shipping_cost

print(f"{'#'*25} FINAL OUTPUT {'#'*25}")
for s_product_det in product_catalogue:
    print(f"{s_product_det}, Quantity: {product_catalogue[s_product_det]['quantity']}, Total Amount: {product_catalogue[s_product_det]['quantity']*product_catalogue[s_product_det]['price']}")
print(f"Total Amount Before Discount : {total_amount_before_dic}")
print(f"Discount_Name : {final_discount_name}, Discount_Amount : {final_discount_amount}")
print(f"Gift Wrap Fee : {gift_wrapping_cost}, Shipping Fee : {total_shipping_cost}")
print(f"Total : {Final_total_bill_amount}")

